build comment boxes and style them

poncem91 opened the pull request:
Makes use of a new Comment java class to store comment's information separately instead of using a string. A helper function was added to the getComments() function to build comment boxes inside the comment history that show each comment's submitter name and the comment's timestamp. If an email was provided, the name in the comment links to the email.

poncem91 committed:
build comment boxes and style them

Commit makes use of a new Comment java class to store comment's information separately instead of using a string. A helper function was added to the getComments() function to build comment boxes inside the comment history that show each comment's submitter name and the comment's timestamp. If an email was provided, the name in the comment links to the email.

poncem91 committed:
Merge branch 'Mafe_AdditionalFeatures_DeleteComments' into Mafe_CommentsDesign

merge changes done in base branch

peap commented on portfolio/src/main/java/com/google/sps/data/Comment.java:
```diff
@@ -0,0 +1,42 @@
+    this.name = name;
+    this.email = email;
+    this.message = message;
+    this.timestamp = this.getTimestamp(timestamp);
```
Because you are converting it from one format to another, I would give the class property `timestamp` a different name than the constructor argument `timestamp`. When they have the same name, it could be easy to confuse them.

Generally, we call these int/long numbers timestamps and the human-readable versions that show a date and a time a "datetime".

poncem91 replied:
"datetime" works so much better for this, I wasn't sure what to call it before but timestamp again. I'll make these changes, thanks!

peap commented on portfolio/src/main/webapp/script.js:
```diff
@@ -119,18 +119,62 @@ document.addEventListener('keydown', function(event) {
 
+
+/** Fetch Comments Helper Function that constructs commentNodes */
+function fetchCommentsHelper(comment) {
```
"Helper" is one of those names that doesn't always provide as much information to readers of the code as it could. (It could be doing anything, if you think about it.) Prefer to call this function something more direct, for example `createCommentNode`.

poncem91 replied:
That makes sense, I'll rename this to be more self-explanatory.

peap commented on portfolio/src/main/webapp/script.js:
```diff
@@ -119,18 +119,62 @@ document.addEventListener('keydown', function(event) {
+    commentNode.classList.add("comment");
+
+    var headerNode = document.createElement('div');
+    headerNode.classList.add("comment-row");
+    headerNode.classList.add("comment-header");
```
`classList.add()` accepts multiple classes as separate arguments, so you can combine these two lines.

poncem91 replied:
Will do!

peap commented on portfolio/src/main/webapp/script.js:
```diff
@@ -119,18 +119,62 @@ document.addEventListener('keydown', function(event) {
-        listNode.innerText = message;
-        commentsHistory.appendChild(listNode);
+
+    comments.forEach(comment => {
+        var commentNode = fetchCommentsHelper(comment);
+        commentsHistory.appendChild(commentNode);
```
This pattern here of applying a function to each element of an array and putting the result into a new array is called mapping. If you want an extra challenge, check out `Array.map()`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map. Because it returns an array, you can call `.map()` on the result, as well, and chain them.

poncem91 replied:
I'll definitely look into this tomorrow! :)

poncem91 replied:
Done! I ended up chaining `.map()` with `forEach()` since I didn't need an array returned on the second portion.

peap reviewed (changes requested):
These comment boxes look great! Nice job.

poncem91 committed:
Merge branch 'Mafe_AdditionalFeatures_DeleteComments' into Mafe_CommentsDesign

merge changes to base branch

poncem91 committed:
rename Comment timestamp property and helper function

Renamed the previously named timestamp property in the Comment class to the better suited name: datetime. Also renamed the fetchCommentsHelper() function to constructCommentNode() to make the name more readable. Consolidated the two headerNode.classList.add() lines into one.

poncem91 committed:
use mapping to construct comment nodes

Makes use of mapping to construct an array of comment nodes from the array of comment objects and then appends each of the nodes into commentsHistory.

poncem91 committed:
change selected maxcomments value in dropdown menu to match new default

Dropdown menu now shows 5 as a default instead of 10.

peap commented on portfolio/src/main/java/com/google/sps/data/Comment.java:
```diff
@@ -23,14 +23,14 @@
     this.email = email;
     this.message = message;
-    this.timestamp = this.getTimestamp(timestamp);
+    this.datetime = this.getTimestamp(timestamp);
```
Let's update the name of this function to match what it returns, as well.

poncem91 replied:
whoops, will update this!

peap commented on portfolio/src/main/webapp/script.js:
```diff
@@ -126,24 +126,21 @@ function getComments(maxComments) {
     const commentsHistory = document.getElementById('comments-history');
     commentsHistory.innerHTML = '';
+    
+    comments.map(comments => constructCommentNode(comments)).forEach(node => commentsHistory.appendChild(node));
```
You can be even more concise here by giving `map` just the name of the function, instead of a lambda. The default behavior is to call the provided function with each element in the first argument position, so if your function matches that signature (like `constructCommentNode` does), you can just give it the name of that function.

So: `comments.map(constructCommentNode)` is the same as the bit that you have now.

poncem91 replied:
Ah, this is the correct syntax for this - I forget this is possible in JS, I had been trying to do `comments.map(constructCommentNode(comments))` but ran into errors. Thanks! I'll update this.

poncem91 committed:
rename Comment class method and be more concise using .map()

Renames Comment class method from getTimestamp() to getDatetime() to align with the property name and makes a small change in the mapping function that constructs comment nodes to make it more concise.
